Esp32 s2 patch2 - debug improvement, and starting to collect some esp32 tools

softhack007 opened the pull request:
### Serial Debug improvement
Lolin S2 mini does not expose the standard RX/TX pins, so it was impossible to get a crashdump via serial-to-usb module.
* Workaround: re-init Serial using pins defined by arduino variant
* use newer Framework (arduino-esp32 2.0.9) that does not block if USB-CDC is unconnected

### esp32Tools
* adding esp32Tools for collecting esp32 specific support functions. The interface in esp32tools.h is generic to allow porting to other platforms. System tools for esp32 are based on WLEDMM source code, and compatible with all esp32 variants.

* The concept could be extended later, to collect all platform specific stuff in a separate "platforms" directory - Similar to fastLED.

* As some crashes happen "silently", add a crash warning to Serial output on startup:

```
Rebooting...
WARNING - possible crash: (code 4) Software reset due to panic or unhandled exception (SW error). Core#0 (code 3) SW reset.
Ready.
```

softhack007 committed:
-S2 debug improvements

Lolin S2 mini does not expose the standard RX/TX pins, so it was impossible to get a crashdump via serial-to-usb module.

* Workaround: re-init Serial using pins defined by arduino variant
* use newer Framework (arduino-esp32 2.0.9) that does not block if USB-CDC is unconnected

softhack007 committed:
add esp32Tools

* adding esp32Tools for collecting esp32 specific support functions. The interface in esp32tools.cpp is generic to allow porting to other platforms.

System tools for esp32 are based on WLEDMM source code, and compatible with all esp32 variants.

Actually the concept could be extended later, to collect all platform specific stuff in a separate "platforms" directory - Similar to fastLED.

* As some crashes happen "silently", add a crash warning to Serial output on startup.

MoonModules commented on src/Sys/esp32Tools.cpp:
```diff
@@ -0,0 +1,250 @@
+    case 13: //13 = RTC Watch dog Reset CPU
+    case 16: //16 = RTC Watch dog reset digital core and rtc module
+    case 17: //17 = Time Group1 reset CPU
+      return F("watchdog"); break;
```
Is F() still needed? 

As:
"const" and literal strings will normally be put in flash memory by default.

?

softhack007 replied:
Your right, will change 👍

MoonModules commented on src/Sys/esp32Tools.cpp:
```diff
@@ -0,0 +1,250 @@
+    }
+  return(F("unknown"));
+}
+
```
See also here: https://github.com/ewowi/StarMod/blob/19da7cb0145f5050920eeea2aab4618bbb0c9bc3/src/Sys/SysModSystem.cpp#L142

Duplicate code, one can be removed, which is the best to remove?

softhack007 replied:
Ah right, I'll try to remove the code in SysModSystem::addResetReasonsSelect.
I'm not much familiar with ArduinoJSON, but let's see.

softhack007 committed:
get rid of all F

not necessary on esp32 - all constant strings are automatically put into flash.

softhack007 committed:
SysModSystem UI text based on esp32Tools

removing duplicate implementation.
